### PR TITLE
*: optimize query statements summary table by avoid decode plan for each row

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve?

Before this PR, query the  `statemens summary` table maybe very slow cause by decode plan, event the query doesn't need to read the `plan`.

### Bench

Generate 3000 statements to filling   `statemens summary` , and each SQL length is `10240 byte`.

1. Use [loadgen](https://github.com/crazycs520/load) to generated load:

   ```shell
   bin/loadgen payload  gen-stmt --stmt-cnt 3000 --query-len 10240 --host 172.16.5.40 --port 4019  --concurrency 10
   ```
   The generated SQL is look like this:
   
   ```sql
   select * from t_gen_stmt_1 where a = 1 or b = 1 or c = 1 
   union all
   select * from t_gen_stmt_2 where a = 2 or b = 2 or c = 2 
   union all
   ...
   ...
   ...
   ```
   
2. Wait generated statements(SQL) fill in the  `statemens summary`.

3. Bench the SQL: `select count(*) from information_schema.cluster_statements_summary_history;`

Bellow is the bench result compare with the master:

|                      | Master | This PR |
| ---------------------| ------ | ------- |
| Query duration       | 3.4s   | 7ms     |
| Query memory consume | 500MB  | < 1MB   |

Related TiDB memory consume metric is below:

Master:

image

This PR:

image

## # What is changed and how it works?

This PR add a special `stmtSummaryReader` to read the  `statemens summary` table data, and only `'decode'` the needed column, avoid decode plan for each table row.

Also, this PR implements `prune columns` for `LogicalMemTable` if the `mem table` is   `statemens summary` table. Otherwise, `stmtSummaryReader` still has to read all column since `LogicalMemTable` doesn't `prune column`.

### Related changes

- N/A

### Check List 

Tests 

- Unit test
- Manual test

Side effects

- - N/A

### Release note 

- N/A

### CODE REVIEW

**Comment by crazycs520** [comment]
@djshow832 @winoros @ClSlaid PTAL

File: util/stmtsummary/reader.go [hunk 1]
```diff
@@ -0,0 +1,557 @@
...
+	}
+	// initialize column value factories.
+	reader.columnValueFactories = make([]columnValueFactory, len(reader.columns))
+	for i, col := range reader.columns {
+		factory, ok := columnValueFactoryMap[col.Name.O]
+		if !ok {
+			if col.Name.O == util.ClusterTableInstanceColumnName {
+				factory = func(ssElement *stmtSummaryByDigestElement, ssbd *stmtSummaryByDigest) interface{} {
+					return reader.instanceAddr
+				}
```
Comment (ClSlaid) [inline util/stmtsummary/reader.go:58]: What is this function list used for? a map taking 2 parameters?

Comment (crazycs520) [reply util/stmtsummary/reader.go:58]: This is used to set the `INSTANCE` column value, for the cluster_xxx table.

File: planner/core/rule_column_pruning.go [hunk 2]
```diff
@@ -281,6 +282,27 @@ func (ds *DataSource) PruneColumns(parentUsedCols []*expression.Column) error {
...
+		infoschema.ClusterTableStatementsSummaryHistory:
+		// currently prune mem-table column only use for statements summary table.
+	default:
+		return nil
+	}
+	used := expression.GetUsedList(parentUsedCols, p.schema)
+	for i := len(used) - 1; i >= 0; i-- {
+		if !used[i] && p.schema.Len() > 1 {
+			p.schema.Columns = append(p.schema.Columns[:i], p.schema.Columns[i+1:]...)
+			p.Columns = append(p.Columns[:i], p.Columns[i+1:]...)
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply planner/core/rule_column_pruning.go:300]: The empty list need to allocate new memory, I like the current approach.

File: executor/infoschema_reader.go [hunk 3]
```diff
@@ -143,13 +143,9 @@ func (e *memtableRetriever) retrieve(ctx context.Context, sctx sessionctx.Contex
...
 			err = e.dataForTiKVStoreStatus(sctx)
-		case infoschema.TableStatementsSummary,
-			infoschema.TableStatementsSummaryHistory,
-			infoschema.TableStatementsSummaryEvicted,
-			infoschema.ClusterTableStatementsSummary,
-			infoschema.ClusterTableStatementsSummaryHistory,
+		case infoschema.TableStatementsSummaryEvicted,
 			infoschema.ClusterTableStatementsSummaryEvicted:
-			err = e.setDataForStatementsSummary(sctx, e.table.Name.O)
+			err = e.setDataForStatementsSummary(sctx)
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply executor/infoschema_reader.go]: done.

**Review by djshow832 — APPROVED** [review]

**Comment by ClSlaid** [comment]
LGTM

**Review by ClSlaid — APPROVED** [review]
LGTM

**Review by AilinKid — APPROVED** [review]

**Pull request merged**